diff --git a/winpr/libwinpr/CMakeLists.txt b/winpr/libwinpr/CMakeLists.txt
index d9673f2d4..5aa37b606 100644
--- a/winpr/libwinpr/CMakeLists.txt
+++ b/winpr/libwinpr/CMakeLists.txt
@@ -83,8 +83,10 @@ check_function_exists(timer_delete TIMER_DELETE)
 check_function_exists(timer_settime TIMER_SETTIME)
 check_function_exists(timer_gettime TIMER_GETTIME)
 if (TIMER_CREATE AND TIMER_DELETE AND TIMER_SETTIME AND TIMER_GETTIME)
-    add_definitions(-DWITH_POSIX_TIMER)
-	winpr_library_add_private(rt)
+    if(NOT IOS)
+        add_definitions(-DWITH_POSIX_TIMER)
+        winpr_library_add(rt)
+    endif()
 endif()
 
 if (ANDROID)
