diff --git a/cmake/FindOpenSSL.cmake b/cmake/FindOpenSSL.cmake
index 11874e085..73fc89c8d 100644
--- a/cmake/FindOpenSSL.cmake
+++ b/cmake/FindOpenSSL.cmake
@@ -71,7 +71,7 @@ IF(MSVC)
   endif()
 ENDIF(MSVC)
 
-IF(ANDROID)
+IF(ANDROID AND WITH_FREERDP_OPENSSL)
     FIND_LIBRARY(OPENSSL_LIBRARIES
       NAMES
         "freerdp-openssl"
@@ -254,7 +254,7 @@ ELSE(WIN32 AND NOT CYGWIN)
 
   SET(OPENSSL_LIBRARIES ${OPENSSL_SSL_LIBRARY} ${OPENSSL_CRYPTO_LIBRARY})
 
-ENDIF(ANDROID)
+ENDIF()
 
 function(from_hex HEX DEC)
   string(TOUPPER "${HEX}" HEX)
