diff --git a/CMakeLists.txt b/CMakeLists.txt
index 312081bfe..cb45dcdb2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -256,6 +256,9 @@ if(CMAKE_COMPILER_IS_GNUCC)
 		else()
 			set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=i686")
 		endif()
+		if(ANDROID)
+			set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--no-warn-shared-textrel")
+		endif()
 	else()
 		if(CMAKE_POSITION_INDEPENDENT_CODE)
 			if(${CMAKE_VERSION} VERSION_LESS 2.8.9)
