diff --git a/CMakeLists.txt b/CMakeLists.txt
index fed7c8a85..d190f1be7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -602,17 +602,9 @@ if(NOT IOS)
 endif()
 
 if(NOT WIN32)
-	check_symbol_exists(pthread_mutex_timedlock pthread.h HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL)
-	if(NOT HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL)
-		check_library_exists(pthread pthread_mutex_timedlock "" HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB)
-	endif()
-	if(NOT HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB)
-		check_library_exists(pthreads pthread_mutex_timedlock "" HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIBS)
-		set(HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB ${HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIBS})
-	endif()
-	if(HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL AND HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB)
-		set(HAVE_PTHREAD_MUTEX_TIMEDLOCK 1)
-	endif()
+	list(APPEND CMAKE_REQUIRED_LIBRARIES pthread)
+	check_symbol_exists(pthread_mutex_timedlock pthread.h HAVE_PTHREAD_MUTEX_TIMEDLOCK)
+	list(REMOVE_ITEM CMAKE_REQUIRED_LIBRARIES pthread)
 endif()
 
 if(WITH_VALGRIND_MEMCHECK)
