diff --git a/cmake/FindOpenSLES.cmake b/cmake/FindOpenSLES.cmake
index 5bbd5e172..480e5f163 100644
--- a/cmake/FindOpenSLES.cmake
+++ b/cmake/FindOpenSLES.cmake
@@ -17,7 +17,8 @@ find_path(OPENSLES_INCLUDE_DIR SLES/OpenSLES.h
 	HINTS ${PC_OPENSLES_INCLUDE_DIR})
 
 find_library(OPENSLES_LIBRARY NAMES OpenSLES
-	HINTS ${PC_OPENSLES_LIBDIR} ${PC_OPENSLES_LIBRARY_DIRS})
+	HINTS ${PC_OPENSLES_LIBDIR} ${PC_OPENSLES_LIBRARY_DIRS}
+	PATH_SUFFIXES "lib64")
 
 # Handle the QUIETLY and REQUIRED arguments and set OPENSL_FOUND to TRUE if
 # all listed variables are TRUE.
