diff --git a/winpr/CMakeLists.txt b/winpr/CMakeLists.txt
index a41d3ed43..5a7b380a9 100644
--- a/winpr/CMakeLists.txt
+++ b/winpr/CMakeLists.txt
@@ -194,7 +194,15 @@ add_subdirectory(include)
 
 add_subdirectory(libwinpr)
 
-if(NOT ANDROID AND NOT IOS AND NOT UWP AND WITH_WINPR_TOOLS)
+if(NOT DEFINED WITH_WINPR_TOOLS)
+	if(NOT ANDROID AND NOT IOS AND NOT UWP)
+		set(WITH_WINPR_TOOLS 0)
+	else()
+		set(WITH_WINPR_TOOLS 1)
+	endif()
+endif()
+
+if(WITH_WINPR_TOOLS)
 	add_subdirectory(tools)
 endif()
 
