diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2ab18fce2..df489e38e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -108,6 +108,21 @@ message("FREERDP_VERSION=${FREERDP_VERSION_FULL}")
 
 set(FREERDP_INCLUDE_DIR "include/freerdp${FREERDP_VERSION_MAJOR}/")
 
+if(NOT DEFINED WITH_FREERDP)
+	set(WITH_FREERDP ON)
+endif()
+
+if(WINPR_ONLY)
+	set(WITH_FREERDP OFF)
+	set(WITH_CHANNELS OFF)
+	set(WITH_CLIENT OFF)
+	set(WITH_SERVER OFF)
+	set(WITH_RDTK OFF)
+	set(WITH_DOTNET OFF)
+	set(WITH_WAYLAND OFF)
+	set(WITH_CLIENT_COMMON OFF)
+endif()
+
 # Compatibility options
 if(DEFINED STATIC_CHANNELS)
 	message(WARNING "STATIC_CHANNELS is obsolete, please use BUILTIN_CHANNELS instead")
@@ -1030,9 +1045,11 @@ if(WITH_THIRD_PARTY)
 	endif()
 endif()
 
-add_subdirectory(include)
+if(WITH_FREERDP)
+	add_subdirectory(include)
 
-add_subdirectory(libfreerdp)
+	add_subdirectory(libfreerdp)
+endif()
 
 if (IOS AND NOT DEFINED CMAKE_OSX_SYSROOT)
 	set(CMAKE_OSX_DEPLOYMENT_TARGET "10.0")
@@ -1050,7 +1067,7 @@ if(WITH_RDTK)
 	add_subdirectory(rdtk)
 endif()
 
-if(WAYLAND_FOUND)
+if(WITH_WAYLAND AND WAYLAND_FOUND)
 	add_subdirectory(uwac)
 endif()
 
@@ -1071,7 +1088,7 @@ if(WITH_CHANNELS)
 endif()
 
 if(WITH_CLIENT_COMMON OR WITH_CLIENT)
-add_subdirectory(client)
+	add_subdirectory(client)
 endif()
 
 if(WITH_SERVER)
diff --git a/winpr/libwinpr/sspi/NTLM/ntlm.c b/winpr/libwinpr/sspi/NTLM/ntlm.c
index 65a5f692f..cde99f77a 100644
--- a/winpr/libwinpr/sspi/NTLM/ntlm.c
+++ b/winpr/libwinpr/sspi/NTLM/ntlm.c
@@ -28,7 +28,6 @@
 #include <winpr/sysinfo.h>
 #include <winpr/registry.h>
 #include <winpr/endian.h>
-#include <freerdp/build-config.h>
 
 #include "ntlm.h"
 #include "../sspi.h"
@@ -38,7 +37,7 @@
 #include "../../log.h"
 #define TAG WINPR_TAG("sspi.NTLM")
 
-#define WINPR_KEY "Software\\" FREERDP_VENDOR_STRING "\\" FREERDP_PRODUCT_STRING "\\WinPR\\NTLM"
+#define WINPR_KEY "Software\\WinPR\\NTLM"
 
 static const char* NTLM_PACKAGE_NAME = "NTLM";
 
